File: app/src/main/java/com/vsa/filmoteca/presenter/widget/EventsWidgetPresenterImpl.java
package com.vsa.filmoteca.presenter.widget;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;

import com.loopj.android.http.AsyncHttpClient;
import com.loopj.android.http.AsyncHttpResponseHandler;
import com.vsa.filmoteca.model.Movie;
import com.vsa.filmoteca.model.MoviesFactory;
import com.vsa.filmoteca.model.database.WidgetDataSource;
import com.vsa.filmoteca.model.sharedpreferences.SharedPreferencesManager;
import com.vsa.filmoteca.presenter.utils.Constants;
import com.vsa.filmoteca.view.EventsWidgetView;

import java.util.List;

import cz.msebera.android.httpclient.Header;

/**
 * Created by seldon on 27/03/15.
 */
public class EventsWidgetPresenterImpl extends AsyncHttpResponseHandler implements EventsWidgetPresenter {

    private EventsWidgetView mView;

    private int mCurrentMovieIndex = 0;
    private int mMoviesListSize = 0;

    private List<Movie> mMovies;

    private Context mContext;

    public EventsWidgetPresenterImpl(EventsWidgetView view) {
        mView = view;
    }

    @Override
    public void onSuccess(int statusCode, Header[] headers, byte[] responseBody) {
        mMovies = MoviesFactory.parseMoviesList(new String(responseBody));
        if (mMovies == null) {
            mView.showRefreshButton(mContext);
        } else {
            if (mMovies.size() != 0) {
                //Actualizando base de datos
                WidgetDataSource widgetDataSource = new WidgetDataSource(mContext);
                widgetDataSource.open();
                widgetDataSource.clearMovies();
                for (int x = 0; x < mMovies.size(); x++)
                    widgetDataSource.insertMovie(x, mMovies.get(x));
                widgetDataSource.close();

                //Estableciendo elemento actual y tamaño de la base de datos
                mCurrentMovieIndex = 0;
                mMoviesListSize = mMovies.size();
                SharedPreferencesManager.setCurrentMovieIndex(mContext, 0);
                SharedPreferencesManager.setMoviesCount(mContext, mMoviesListSize);

                //Configurando la vista
                mView.setupLRButtons(mContext);
                Movie movie = null;
                if (mMovies != null && mMovies.size() > 0)
                    movie = mMovies.get(mCurrentMovieIndex);
                mView.hideProgress();
                mView.setupMovieView(mContext, movie);
                mView.setupIndexView(mContext, mCurrentMovieIndex + 1, mMoviesListSize);
                mView.updateWidget();
            }
        }
    }

    @Override
    public void onFailure(int statusCode, Header[] headers, byte[] responseBody, Throwable error) {

    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        mContext = context;
        mView.initWidget(context);
        mView.showProgress();
        AsyncHttpClient client = new AsyncHttpClient();
        client.setTimeout(Constants.TIMEOUT_WIDGET);
        client.get(Constants.URL_SOURCE, this);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        mContext = context;
        mView.initWidget(mContext);
        if (intent.getAction().equals(Constants.ACTION_WIDGET_LEFT) || intent.getAction().equals(Constants.ACTION_WIDGET_RIGHT)) {
            //Obtenemos el indice actual y el tamaño de la base de datos
            mCurrentMovieIndex = SharedPreferencesManager.getCurrentMovieIndex(context);
            mMoviesListSize = SharedPreferencesManager.getMoviesCount(context);
            if (mMoviesListSize != 0) {
                if (intent.getAction().equals(Constants.ACTION_WIDGET_LEFT)) {
                    if (mCurrentMovieIndex > 0) {
                        mCurrentMovieIndex--;
                        //Actualizamos el valor del indice
                        SharedPreferencesManager.setCurrentMovieIndex(context, mCurrentMovieIndex);
                    }
                } else if (intent.getAction().equals(Constants.ACTION_WIDGET_RIGHT)) {
                    if (mCurrentMovieIndex < (mMoviesListSize - 1)) {
                        mCurrentMovieIndex++;
                        //Actualizamos el valor del indice
                        SharedPreferencesManager.setCurrentMovieIndex(context, mCurrentMovieIndex);
                    }
                }
                //Get the movie from database
                WidgetDataSource widgetDataSource = new WidgetDataSource(context);
                widgetDataSource.open();
                Movie movie = widgetDataSource.getMovie(mCurrentMovieIndex);
                widgetDataSource.close();
                //Preparamos la vista
                mView.setupLRButtons(context);
                mView.setupMovieView(context, movie);
                mView.setupIndexView(context, mCurrentMovieIndex + 1, mMoviesListSize);
                mView.updateWidget();
            }
        }
    }

    @Override
    public void loadMovies() {

    }
}

File: app/src/main/java/com/vsa/filmoteca/presentation/presenter/widget/EventsWidgetPresenterImpl.kt
package com.vsa.filmoteca.presentation.presenter.widget

import com.vsa.filmoteca.domain.model.Movie
import com.vsa.filmoteca.domain.usecase.GetMoviesListUseCase
import com.vsa.filmoteca.domain.usecase.MoviesPersistanceUseCase
import com.vsa.filmoteca.presentation.view.EventsWidgetView
import javax.inject.Inject

/**
 * Created by seldon on 27/03/15.
 */
class EventsWidgetPresenterImpl @Inject constructor(
    private val view: EventsWidgetView,
    private val getMoviesListUseCase: GetMoviesListUseCase,
    private val moviesPersistanceUseCase: MoviesPersistanceUseCase
) : EventsWidgetPresenter {

    private var currentMovieIndex = 0
    private var moviesListSize = 0

    override fun onRefresh() {
        view.showProgress()
        view.refreshViews()
        requestMoviesList()
    }

    override fun onButtonLeftClick() {
        //Obtenemos el indice actual y el tamaño de la base de datos
        currentMovieIndex = moviesPersistanceUseCase.currentMovieIndex
        moviesListSize = moviesPersistanceUseCase.moviesCount
        if (moviesListSize != 0) {
            if (currentMovieIndex > 0) {
                currentMovieIndex--
                //Actualizamos el valor del indice
                moviesPersistanceUseCase.currentMovieIndex = currentMovieIndex
            }
            val movie = moviesPersistanceUseCase.currentMovie
            //Preparamos la vista
            view.setupLRButtons()
            view.setupMovieView(movie.url, movie.title, movie.date)
            view.setupIndexView(currentMovieIndex + 1, moviesListSize)
            view.refreshViews()
        }
    }

    override fun onButtonRightClick() {
        //Obtenemos el indice actual y el tamaño de la base de datos
        currentMovieIndex = moviesPersistanceUseCase.currentMovieIndex
        moviesListSize = moviesPersistanceUseCase.moviesCount
        if (moviesListSize != 0) {

            if (currentMovieIndex < moviesListSize - 1) {
                currentMovieIndex++
                //Actualizamos el valor del indice
                moviesPersistanceUseCase.currentMovieIndex = currentMovieIndex
            }
            val movie = moviesPersistanceUseCase.currentMovie
            //Preparamos la vista
            view.setupLRButtons()
            view.setupMovieView(movie.url, movie.title, movie.date)
            view.setupIndexView(currentMovieIndex + 1, moviesListSize)
            view.refreshViews()
        }
    }

    private fun requestMoviesList() {
        getMoviesListUseCase.moviesList {
            it.fold({ movies ->
                if (movies.isEmpty()) {
                    view.showRefreshButton()
                } else {
                    if (movies.isNotEmpty()) {
                        //Actualizando base de datos
                        moviesPersistanceUseCase.setMovies(movies)

                        //Estableciendo elemento actual y tamaño de la base de datos
                        currentMovieIndex = 0
                        moviesListSize = movies.size
                        moviesPersistanceUseCase.currentMovieIndex = 0
                        moviesPersistanceUseCase.moviesCount = moviesListSize

                        //Configurando la vista
                        view.setupLRButtons()
                        var currentMovie: Movie? = null
                        if (movies.isNotEmpty()) {
                            currentMovie = movies[currentMovieIndex]
                            view.setupMovieView(
                                currentMovie.url,
                                currentMovie.title,
                                currentMovie.date
                            )
                            view.setupIndexView(currentMovieIndex + 1, moviesListSize)
                        }
                        view.hideProgress()
                    }
                }
                view.refreshViews()
            }, {
                view.showRefreshButton()
                view.refreshViews()
            })
        }
    }

}
